ide_extensions.md

```diff
@@ -0,0 +1,161 @@
+
+## Language Server Protocol (LSP) Deep Dive
+LSP is a relatively recent standard (the first public release was in 2017) for a complex topic (programming language analysis).
+As a result, the ecosystem is a bit chaotic: there are often multiple competing LSP servers for a given language and high-level documentation can be sparse.
+
+Luckily, the [LSP specification][lsp-spec] itself is extremely readable, and the JSON-RPC2 protocol makes introspection human-friendly.
+
+### Go Support
+The [go.lsp.dev][] project provides a JSON-RPC2 server and client implementation in addition to Go structs for the LSP wire messages.
+It's an exported version of the inaccessible implementation used for the official Go LSP server (`gopls`) from [golang.org/x/tools/internal/lsp][x-tools-lsp].
+Note: it appears to have diverged at this point; stewardship and affiliation (if any) to Go team is unclear as of February 2022.
+
+Developing an LSP in Go is a natural choice:
+ * Possible to integrate directly into Tilt (`tilt lsp` command) eliminating extra dependencies
+ * Tilt team familiarity with Go
+ * Official Starlark parser implementation exists
+   * See caveats in [Starlark Parsing](#starlark-parsing) section below
+
+### Starlark Parsing
+While writing code, it's very normal for the file to be syntactically invalid.
+This is often at odds with language parsers, which strictly follow a formal grammar.
+
+Microsoft wrote a fantastic [overview of lessons learned][fault-tolerant-parser] from writing a fault tolerant PHP parser for use in VSCode.
+
+#### starlark-go
+The [starlark-go][] parser stops at the first error it encounters.
+Furthermore, it does this via calling `recover()` from a `panic()` rather than more idiomatic Go error-handling.
+
+That said, the starlark-go parsing code is hand-written and roughly ~1000 LOC, so creating a fault-tolerant version is not insurmountable.
+It's unlikely we'd be able to upstream this, but the Starlark language spec is a slow-moving target, so the ongoing maintenance burden here would be minimal.
+
+#### Tree-sitter
+Another approach worth exploring is to generate a [Tree-sitter][] grammar.
+Tree-sitter is designed for fast parsing, incremental updates, and graceful error recovery (all important properties for an LSP server).
+It's written in dependency-free C, which makes bindings for a variety of languages practical including Go and WASM.
+
+Microsoft even uses it in the [vscode-anycode][] extension, which provides generic LSP features using _any_ Tree-sitter AST.
+
+Tree-sitter also has highlighting functionality that can be used to provide semantic syntax highlighting that's more accurate than the TextMate-based regex approach.
+(This would also be useful for a future IntelliJ extension: see the notes in the [Syntax Highlighting](#syntax-highlighting) section.)
+
+### Conclusion
+We should write our LSP in Go.
+The LSP Go packages, while under-documented, are feature complete and real world tested via `gopls`.
```

Comment:
Ooo, yes, this is a great point - I've definitely heard horror stories about `gopls` as well!

I _believe_ most of the issues have come from the "Go analysis" part of `gopls` rather than the protocol (LSP+JSONRPC2) pieces so shouldn't impact us, but I'll take a quick peek at public issues/bugfixes for it to verify that.